File: HackerRank/Interview Preparation Kit/Linked Lists/DetectACycle.java
/*
Detect a cycle in a linked list. Note that the head pointer may be 'null' if the list is empty.

A Node is defined as: 
    class Node {
        int data;
        Node next;
    }
*/

boolean hasCycle(Node head) {
    if (head == null) return false;
    
    Node currNode = head;
    Set<Node> seen = new HashSet<>();
    while (currNode != null) {
        seen.add(currNode);
        currNode = currNode.next;
        if (seen.contains(currNode)) return true;
    }
    return false;
}

File: HackerRank/Interview Preparation Kit/Linked Lists/detect_a_cycle.cpp
/*
Detect a cycle in a linked list. Note that the head pointer may be 'NULL' if the list is empty.

A Node is defined as: 
    struct Node {
        int data;
        struct Node* next;
    }
*/

bool has_cycle(Node* head) {
    // Complete this function
    // Do not write the main method
    Node* fast = head;
    Node* slow = head;
    while(fast != NULL && slow != NULL && fast->next) {
        fast = fast->next->next;
        slow = slow->next;
        if(fast == slow) {
            return true;
        }
    }
    
    return false;
    
}